cmd/snap/cmd_routine_portal_info.go

```diff
@@ -0,0 +1,148 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+	"text/template"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/sandbox/cgroup"
+)
+
+type cmdRoutinePortalInfo struct {
+	clientMixin
+	PortalInfoOptions struct {
+		Pid int
+	} `positional-args:"true" required:"true"`
+}
+
+var shortRoutinePortalInfoHelp = i18n.G("Return information about a process")
+var longRoutinePortalInfoHelp = i18n.G(`
+The portal-info command returns information about a process in keyfile format.
+
+This command is used by the xdg-desktop-portal service to retrieve
+information about snap confined processes.
+`)
+
+func init() {
+	addRoutineCommand("portal-info", shortRoutinePortalInfoHelp, longRoutinePortalInfoHelp, func() flags.Commander {
+		return &cmdRoutinePortalInfo{}
+	}, nil, []argDesc{{
+		// TRANSLATORS: This needs to begin with < and end with >
+		name: i18n.G("<process ID>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Process ID of confined app"),
+	}})
+}
+
+const portalInfoTemplate = `[Snap Info]
+InstanceName={{.Snap.Name}}
+{{- if .App}}
+AppName={{.App.Name}}
+{{- end}}
+{{- if .DesktopFile}}
+DesktopFile={{.DesktopFile}}
+{{- end}}
+HasNetwork={{.HasNetwork}}
+`
+
+var (
+	cgroupSnapNameFromPid   = cgroup.SnapNameFromPid
+	apparmorSnapNameFromPid = apparmor.SnapNameFromPid
+)
+
+func (x *cmdRoutinePortalInfo) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	snapName, err := cgroupSnapNameFromPid(x.PortalInfoOptions.Pid)
+	if err != nil {
+		return err
+	}
+	snap, _, err := x.client.Snap(snapName)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve info for snap %q: %v", snapName, err)
+	}
+
+	// Try to identify the application name from AppArmor
+	var app *client.AppInfo
+	if snapName, appName, _, err := apparmorSnapNameFromPid(x.PortalInfoOptions.Pid); err == nil && snapName == snap.Name && appName != "" {
+		for i := range snap.Apps {
+			if snap.Apps[i].Name == appName {
+				app = &snap.Apps[i]
+				break
+			}
+		}
+	}
+	// As a fallback, pick an app with a desktop file, favouring
+	// the app named identically to the snap.
+	if app == nil {
+		for i := range snap.Apps {
+			if snap.Apps[i].DesktopFile != "" && (app == nil || snap.Apps[i].Name == snap.Name) {
+				app = &snap.Apps[i]
+			}
+		}
+	}
+
+	var desktopFile string
+	if app != nil {
+		desktopFile = filepath.Base(app.DesktopFile)
+	}
+
+	// Determine whether the snap has access to the network
+	connections, err := x.client.Connections(&client.ConnectionOptions{
+		Snap:      snap.Name,
+		Interface: "network",
+	})
+	if err != nil {
+		return fmt.Errorf("cannot get connections for snap %q: %v", snap.Name, err)
+	}
+	var hasNetwork bool
+	for _, conn := range connections.Established {
```

Comment:
Added a comment here

Reply:
@jhenstridge can you comment on this?

Reply:
This is intended to control access to the following portal API, which is 

https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.portal.NetworkMonitor.xml

At present, it will error out if a flatpak without network access attempts to call these APIs.  I want to be able to do the same for snaps.

The portal API seems roughly equivalent to the Unity specific APIs exposed by the `network-status` interface (which I think was implemented by indicator-network?), but I'm not sure whether it makes sense to try and split this out from `network` again.

Reply:
@jhenstridge we do have a network-status interface (related to  com.ubuntu.connectivity1.NetworkingStatus) , maybe this behavior should be folded into that?  @jdstrand any opinion on this?

Reply:
@pedronis - we have the network-status interface that was designed for exactly this use case. In fact, it was designed with the understanding that a future (now today! :) portals API might be added to it. By all means, lets include this access there (it would be nice to have something in there that is actually useful to people since com.ubuntu.connectivity1.NetworkingStatus comes from unity8 APIs).

To @jhenstridge's point, I appreciate wanting to put this in the network interface since that seems to mirror how flatpaks are intended to be designed. Our interfaces design and permissions model is different from flatpak's though so I'd prefer to keep our interfaces as clean as possible and take this to the snapcraft/part/etc level (ie, adjust the templates to include 'network-status' alongside 'network' for gtk3/desktop/etc apps). While somewhat contrived and portals aside, it is not outside of the realm of possibility for a snap to want to have one app check for network status and another app to use the network and since we already have a design for this, IMHO we should stay on that path.

Reply:
I didn't really consider the `network-status` interface, since it looked like a hold over from Unity 8 and looks to have never been used:
* The interface description seems to indicate that it is for a specific `NetworkingStatus` service.
* The interface is not implicit, so could only have been used if another snap provided a matching slot (presumably whatever snap packaged indicator-network).
* It's not clear this ever worked: the plug and slot rules grant access to communicate on the system bus, while indicator-network exposes the interface on the session bus.

So it isn't a simple case of running s/network/network-status/ over the PR.

I guess it is worth asking whether a separate interface makes sense if we don't have a snap providing a slot to connect to.  A snap with the `network` plug can already discover most of the information the portal interface provides, albeit not in a user or battery friendly way (e.g. repeatedly trying to make connections in a busy loop until the network connects).

Reply:
@jhenstridge @jdstrand  either way we'll need to add apparmor dbus access to NetworkMonitor somewhere, no?

we could refine the description of network-status, add it there, and make it implicit and auto-connected on classic for a start? (as we have other interfaces with that pattern)

Reply:
The name of the interface (network-status) was chosen because it was generic so that we may add to its rules as needed in the future. You are right that currently it requires a slot implementation because there wasn't an installed service in classic distro or UC that could make it implied, but we have other interfaces that may be provided by either.

That said, what is different here is that the current implementation intends to give access to com.ubuntu.connectivity1.NetworkingStatus, but that service is defunct and it is not expected to be implemented by any slot. While we have interfaces where a slotting snap is expected to provide multiple dbus interfaces, those are usually a group of related services (and, interestingly, also focused on unity8). In this case, the services are related, but by function (ie, different software that answers the same question), not implementation (a group of dbus services provided by one set of software).

While unusual, in this particular case, I think I'd prefer rather than relegating a great interface name like `network-status` to the annals of history that we add implied portals rules to this interface, updating the comments and documenting as needed. Perhaps the implementation should be: if implied, portals rules, if slot-provided, com.ubuntu.connectivity1.NetworkingStatus rules (with an optional slot interface attribute that allows the slot to declare it is providing just portals or both portals and com.ubuntu.connectivity1.NetworkingStatus).

I see arguments for not doing that (though realistically they ensure we'll never see `network-status` used anywhere) and defer to @pedronis. I will say, if this is the route, it should be in a separate PR.

Reply:
> I see arguments for not doing that (though realistically they ensure we'll never see `network-status` used anywhere) and defer to @pedronis. I will say, if this is the route, it should be in a separate PR.

Unless it is something very simple like @pedronis' suggestion. If we want interface attributes/etc, probably best to review those in isolation.

Reply:
@jdstrand @jhenstridge I definitely prefer that we use an explicit network-status to mark the interest in checking connectivity, vs generic network that can imply only local or not network operations.

The agreement with @jhenstridge is that this can tentatively land once the s/network/network-status/ in the output.

@jdstrand @jhenstridge we need to see how to organize, who can work on a separate PR to rework network-status ?

Reply:
we agreed today with @jhenstridge that he will propose a PR to clean out network-status as first step (it never autoconnected and needs a app-provided slot anyway ATM)